src/misc/ssh/ssh_decode_sequence_multi.c

```diff
@@ -0,0 +1,136 @@
+ */
+#include "tomcrypt_private.h"
+#include <stdarg.h>
+
+/**
+   @file ssh_decode_sequence_multi.c
+   SSH data type representation as per RFC4251, Russ Williams
+*/
+
+#ifdef LTC_SSH
+
+/**
+  Decode a SSH sequence using a VA list
+  @param out    [out] Destination for data
+  @param outlen [in/out] Length of buffer and resulting length of output
+  @remark <...> is of the form <type, data> (int, void*) except for string <type, data, size>
+  @return CRYPT_OK on success
+*/
+int ssh_decode_sequence_multi(const unsigned char *in, unsigned long inlen, ...)
+{
+   int           err;
+   ssh_data_type type;
+   void          *data;
+   va_list       args;
+   unsigned long size, bufsize;
+
+   LTC_ARGCHK(in    != NULL);
+
+   /* Decode values from buffer */
+   va_start(args, inlen);
+   for (;;) {
+      type = (ssh_data_type)va_arg(args, int);
+      data = va_arg(args, void*);
+      LTC_UNUSED_PARAM(data);
+
+      if (type == LTC_SSHDATA_EOL) {
+         break;
+      }
+
+      switch (type) {
+         case LTC_SSHDATA_BYTE:
+            size = ((unsigned long)in[0]);
+            in++;
+            *(unsigned long *)data = size;
+            break;
+         case LTC_SSHDATA_BOOLEAN:
+            size = ((unsigned long)in[0]);
+            in++;
+            /*
+               The value 0 represents FALSE, and the value 1 represents TRUE.  All non-zero values MUST be
+               interpreted as TRUE; however, applications MUST NOT store values other than 0 and 1.
+            */
+            *(unsigned long *)data = (size)?1:0;
+            break;
+         case LTC_SSHDATA_UINT32:
+            size = ((unsigned long)in[0] << 24) |
+                   ((unsigned long)in[1] << 16) |
+                   ((unsigned long)in[2] << 8) |
+                   ((unsigned long)in[3]);
+            in += 4;
+            *(unsigned long *)data = size;
+            break;
+         case LTC_SSHDATA_UINT64:
+            size = ((unsigned long)in[0] << 56) |
+                   ((unsigned long)in[1] << 48) |
+                   ((unsigned long)in[2] << 40) |
+                   ((unsigned long)in[3] << 32) |
+                   ((unsigned long)in[4] << 24) |
+                   ((unsigned long)in[5] << 16) |
+                   ((unsigned long)in[6] << 8) |
+                   ((unsigned long)in[7]);
+            in += 8;
+            *(unsigned long *)data = size;
+            break;
+         case LTC_SSHDATA_STRING:
+         case LTC_SSHDATA_NAMELIST:
+            bufsize = va_arg(args, unsigned long);
+            size = ((unsigned long)in[0] << 24) |
+                   ((unsigned long)in[1] << 16) |
+                   ((unsigned long)in[2] << 8) |
+                   ((unsigned long)in[3]);
+            in += 4;
+            if (size > 0) {
+               if (size >= bufsize) {
+                  va_end(args);
+                  return CRYPT_BUFFER_OVERFLOW;
+               }
+               strncpy((char *)data, (const char *)in, size);
```

Comment:
First I thought why we can't use `strlcpy()`... then I remembered this standardization thingy... as we stay with this please introduce a macro `XSTRNCPY` in `tomcrypt_custom.h` (and also add it to the list in `crypt.tex` and `helper.pl` please)

Reply:
Done. I've also added a check for `strcpy` in `helper.pl` - not used anywhere in the code as far as I can tell, but presumably it should be alerted if someone introduces it.